nodePackages: cleanup overrides and add meta.mainProgram to many packages

malob opened the pull request:
###### Description of changes

Continuing on my adventure to get `nix run` working with more packages in `nixpkgs` (previously #171602), my hacky script identified a bunch of packages in `nodePackages` that were candidates for adding `meta.mainProgram`,  but it couldn't automatically add it due their auto-generated nature. So I went ahead and added them as overrides (2c274c6d28efb56c45c31cc6abd409ef5750e386). I tested them all to make sure they work.

However, I quickly got frustrated by being unsure where to add them, so I went ahead and sorted all the overrides first (4c8844135df26370dde8cceeddca08ece3ba4be3). While doing so, I noticed that there wasn't any consistency between the use of `override` vs `overrideAttrs`, the argument name of in function passed to them (`oldAttrs` vs `attrs` vs `drv`), and that when adding/overriding `meta` it was done in such a way that the old attributes weren't being preserved (as well as fixing a few other small errors). So I went ahead and fixed all of that as well (0b157fcf8fd62add66b0d60d37466b01932358de).

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

malob committed:
nodePackages: sort overrides

malob committed:
nodePackages: cleanup/fix overrides

* Standardize on `override` vs `overrideAttrs'.
* Use `oldAttrs` as standard argument to `override` when it's needed.
* Fix overrides of `meta` so that original attributes are preserved.

malob committed:
nodePackages: make it easy to add meta.mainProgram to packages

This was accomplish by moving the current overrides overlay out of
./default.nix into to a seperate file ./overrides.nix, and composing
that overlay an overlay ,`mainProgramOverrides`, (defined in
./default.nix) which uses the contents of a new file ./main-programs.nix
to create overrides that add `meta.mainProgram` to packages.

The following changes were also made to existing overrides:
* `self` and `super` where changed to `final` and `prev` respectively
* Existing additions of `mainProgram` were moved to ./main-programs.nix
* References to `pkgs.lib` were changed to `lib`
* References to `pkgs.nodejs` were changed to `nodejs`
* References to `nodePackages` were changed  to `final`
* References to `pkgs.callPackage` were changed to `callPackage`

Finally `meta.mainProgram` was added to all packages that provide a
single executable whose name differs from the package's name, for
packages available on `{aarch64,x86_64}-darwin` and `x86_64-linux`.

malob commented:
Result of `nixpkgs-review pr 171863` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>bitwarden-cli</li>
  </ul>
</details>

malob commented on pkgs/development/node-packages/default.nix:
```diff
@@ -40,11 +39,12 @@ let
+    });
 
-    bitwarden-cli = super."@bitwarden/cli".override (drv: {
-      name = "bitwarden-cli-${drv.version}";
```
This is the change that caused the single rebuild.

malob replied:
Where the change was to `name = "bitwarden-cli";`

malob commented:
Just flagging that I added a new commit that vastly simplifies creating overrides just for the purpose of adding `meta.mainProgram`, and moved the existing overrides to use this functionality.

SuperSandro2000 commented on pkgs/development/node-packages/default.nix:
```diff
@@ -67,38 +67,34 @@ let
-        platforms = pkgs.lib.platforms.none;
-      };
+    epgstation = super."epgstation-../../applications/video/epgstation".override (oldAttrs: {
+      meta = oldAttrs.meta // { platforms = pkgs.lib.platforms.none; };
```
I don't understand this line.

malob replied:
I don't understand it either. This was the original (all I did was change `drv` to `oldAttrs` put it all on one line):

```nix
    epgstation = super."epgstation-../../applications/video/epgstation".override (drv: {
      meta = drv.meta // {
        platforms = pkgs.lib.platforms.none;
      };
    });
```

Mindavi replied:
Probably related to comment on L102-103

SuperSandro2000 commented on pkgs/development/node-packages/default.nix:
```diff
@@ -40,11 +39,12 @@ let
-      name = "bitwarden-cli-${drv.version}";
-      meta.mainProgram = "bw";
+    bitwarden-cli = super."@bitwarden/cli".override (oldAttrs: {
+      name = "bitwarden-cli";
```
Can we use pname here?

malob replied:
It seems the `pname` isn't defined, so that's not and option:

```console
~/C/nixpkgs on nodePackages-mainProgram
❯ nix eval -f . nodePackages."@bitwarden/cli".pname
error: attribute 'pname' in selection path 'nodePackages.@bitwarden/cli.pname' not found
       Did you mean name?
```

SuperSandro2000 commented on pkgs/development/node-packages/default.nix:
```diff
@@ -382,16 +378,16 @@ let
+    });
 
-    tailwindcss = super.tailwindcss.overrideAttrs (oldAttrs: {
+    tailwindcss = super.tailwindcss.override {
```
Really unsure about this change.

malob replied:
Honestly I don't understand the subtleties when it comes to using `override` vs `overrideAttrs`, e.g., I was surprised to see `override` being passed a function as an argument, as opposed to an attribute set, in this file.

As far as I can tell this change doesn't produce any meaningful difference to the derivation. `nixpkgs-review` doesn't notice any change that would have it try and rebuild the package, and the package builds and runs fine on my system.

SuperSandro2000 commented on pkgs/development/node-packages/default.nix:
```diff
@@ -8,49 +8,35 @@ let
     inherit (stdenv.hostPlatform) system;
   };
-  self = super // {
+  addMainPrograms = lib.mapAttrs (pkgName: mainProgram:
+    super.${pkgName}.override (oldAttrs: {
+      meta = oldAttrs.meta // { inherit mainProgram; };
+    })
```
Can you merge this with the other commit to make the history less noisy?

malob replied:
Sure!

malob replied:
Done.

Mindavi commented:
I tried building some of the packages with diffs and that seems to be ok.

Note that nodePackages is not recursed into by hydra, ofborg or nixpkgs-review, so it's no surprise there are no rebuilds to me. Even if all of nodePackages was rebuilt none of those tools would report rebuilds.

In general the diff for the first 2 commit looks ok (still need to look at the third), although it's quite messy. But not seeing any big surprises.

Mindavi commented:
Normally override is used to override `arguments` (function call parameters). overrideAttrs is used to override `attributes` (defined inside the derivation). Since it seems like here the `attributes` are changed, `overrideAttrs` _seems_ like the right choice. Not sure how `mkNodeDerivation` (or whatever) looks like though, it might have a different definition. AFAICT the `broken` attribute at least still works even when using `override`.

Mindavi commented:
I wonder, is there any way to not have to do _either_ add a package to the list to easily override the mainProgram, or look if it's already overriden and having to add it there?

Would be really nice to not have that weird split between packages that are in the overrides and aren't.

malob commented:
> I wonder, is there any way to not have to do _either_ add a package to the list to easily override the mainProgram, or look if it's already overriden and having to add it there?
> 
> Would be really nice to not have that weird split between packages that are in the overrides and aren't.

One way to do this would be to compose overlays like they do with `haskellPackages`, and have `default.nix` be a small file that composes two overlays `overrides.nix` which would contain the overrides that are currently in `default.nix`, and `main-program-overrides.nix` that was just used to add `meta.mainProgram` to packages.

Mindavi commented:
Are you willing to make that change in this PR? I do think it's valuable to do that, but I'm not sure if it should be a blocker.

malob commented:
@Mindavi, just pushed the changes, with notes in the commit message. Looks like there's a merge conflict now, so I'll go head and rebase on `master` to address those.

malob commented:
Result of `nixpkgs-review pr 171863` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>bitwarden-cli</li>
    <li>hyperspace-cli</li>
  </ul>
</details>

malob commented:
According to me things look good here now, aside from making some quick tweaks to the documentation which I'll do now.

malob committed:
nodePackages: update/cleanup documentation

malob commented:
@SuperSandro2000, @Mindavi.

Ok, documentation updated (and a fix to my hacky script identified a bunch more packages to add to `main-programs.nix`).

I think this is ready to be merged.

malob commented:
FYI, I'm currently building all packages in `nodePackages` to double check that this PR doesn't introduce any issues, and also to catch any other fixes to overrides that I can add to this PR to improve the state of `nodePackges`.

malob committed:
nodePackages: get more packages building

Packages that now build:
* "@electron-forge/cli
* "@hyperspace/cli"/hyperspace-cli
* "@medable/mdctl-cli"
* thelounge-plugin-giphy
* vega-cli

Also move `pkgs.makeWrapper` from `buildInputs` to `nativeBuildInputs`
in overrides.

malob commented:
@SuperSandro2000, @Mindavi.

Alrighty, I build packages on `{aarch64,x86_64}-darwin` and it doesn't appear that any of the changes to overrides in this PR broke any packages that were previously building. I did identify a bunch of packages that weren't building, and added a commit with fixes (see the comment on the commit for a list).

I wasn't able to quickly fix the following packages:
* `btw-rpc-explorer`: tries to do pull a Git repo that it doesn't have permission to access.
* `gatsby-cli`: missing dependency `node-gyp-build-optional-packages` which need's to be added to `node-packages.json`.
* `parcel`: same as above.
* `readability-cli`: can't patch a shebang that uses `env -S`, and due to how `installPhase` works for Node packages, it's not an easy fix since it's hard to patch the offending file manually.

I'll fix `gatsby-cli` and `parcel` in a separate PR.

Looks like there's another merge conflict now. I'll rebase on `master` again. Then again, I think this is good to merge.

Mindavi reviewed (approved):
Nix is not my first language, but this seems to make sense and it's a lot cleaner than the mix it was before. Thanks for giving nodePacakges some love!

Hoping to get back to this at a later point to have a somewhat better look, but in principle I'm okay if this gets in.

malob commented:
Rebased on `master` to address merge conflicts.

I want to make sure this PR gets the appropriate amount of scrutiny given the complexity of the changes, so I definitely don't want it merged prematurely.

However, I'd like to make a bid that, to the extent to which peoples priorities when it comes to reviewing/merging PR in `nixpkgs` are flexible, this PR get a bump in the priority queue, since basically any change to the old `default.nix` file causes merge conflicts which required non-trivial effort to resolve and ensure the changes get properly propagated through the commits in this PR.

Mindavi commented:
It's not really nice that this'll create merge conflicts, but so be it. I think it's nice to have this working for whenever nix 3 is released (and for everyone using the experimental features already).

Thanks for trying to make things better :smile: .

malob commented:
Thanks!

> It's not really nice that this'll create merge conflicts, but so be it.

FWIW, it will only cause merge conflicts for folks that had an open PR editing `default.nix` before this was merged, and resolving those should be as easy as moving the changes over to `overrides.nix` and renaming `self` and `super` to `final` and `prev` respectively. It was just a little bit more of a hassle on my end since I had to manually hunt for what was changed in `default.nix` when rebasing, and ensure I properly propagated those changes through the commits in this PR.

snawaz commented:
@malob 

> I wasn't able to quickly fix the following packages:
> [..]
> * `gatsby-cli`: missing dependency `node-gyp-build-optional-packages` which need's to be added to `node-packages.json`.
>  [..]
> I'll fix `gatsby-cli` and `parcel` in a separate PR.

Did you get to chance to fix the issue with `node-gyp-build-optional-packages` being missing? 

I'm also facing `node-gyp-build-optional-packages` being `command not found`. It's a transitive dep in my project. So I don't know if there is any fix to this.. 🤔  I'm kinda new to nodejs and nix.

![image](https://user-images.githubusercontent.com/1185906/174096990-5fecf5f8-5434-4323-a72d-fe65ef8f3729.png)

malob commented:
@snawaz, I tried but didn't have any luck unfortunately. I'm not particularly knowledgeable about the the nodejs ecosystem in `nixpkgs` either :(

snawaz commented:
@malob 

Looks like the master now seems to have fixed this: 

> gatsby-cli: missing dependency node-gyp-build-optional-packages which need's to be added to node-packages.json.

I also kinda understand how these things work. I'll try to solve my problem in the same way. 🤔
